lib/matplotlib/axes/_axes.py

```diff
@@ -3398,116 +3398,57 @@ def errorbar(self, x, y, yerr=None, xerr=None,
-                if capsize > 0:
-                    caplines.append(mlines.Line2D(
-                        xo, yo, marker='|', **eb_cap_style))
-            if xuplims.any():
-                xo, yo, lo = apply_mask([x, y, left], xuplims & everymask)
-                if self.xaxis_inverted():
-                    marker = mlines.CARETRIGHTBASE
-                else:
-                    marker = mlines.CARETLEFTBASE
-                caplines.append(mlines.Line2D(
-                    lo, yo, ls='None', marker=marker, **eb_cap_style))
-                if capsize > 0:
-                    caplines.append(mlines.Line2D(
-                        xo, yo, marker='|', **eb_cap_style))
-
-        if yerr is not None:
-            lower, upper = extract_err('y', yerr, y, lolims, uplims)
-            barcols.append(self.vlines(
-                *apply_mask([x, lower, upper], everymask), **eb_lines_style))
-            # select points without upper/lower limits in y and
-            # draw normal errorbars for these points
-            noylims = ~(lolims | uplims)
-            if noylims.any() and capsize > 0:
-                xo, lo, uo = apply_mask([x, lower, upper], noylims & everymask)
-                caplines.extend([
-                    mlines.Line2D(xo, lo, marker='_', **eb_cap_style),
-                    mlines.Line2D(xo, uo, marker='_', **eb_cap_style)])
-            if lolims.any():
-                xo, yo, uo = apply_mask([x, y, upper], lolims & everymask)
-                if self.yaxis_inverted():
-                    marker = mlines.CARETDOWNBASE
-                else:
-                    marker = mlines.CARETUPBASE
-                caplines.append(mlines.Line2D(
-                    xo, uo, ls='None', marker=marker, **eb_cap_style))
-                if capsize > 0:
-                    caplines.append(mlines.Line2D(
-                        xo, yo, marker='_', **eb_cap_style))
-            if uplims.any():
-                xo, yo, lo = apply_mask([x, y, lower], uplims & everymask)
-                if self.yaxis_inverted():
-                    marker = mlines.CARETUPBASE
-                else:
-                    marker = mlines.CARETDOWNBASE
-                caplines.append(mlines.Line2D(
-                    xo, lo, ls='None', marker=marker, **eb_cap_style))
+            low, high = dep + np.row_stack([-(1 - lolims), 1 - uplims]) * err
+
+            barcols.append(lines_func(
+                *apply_mask([indep, low, high], everymask), **eb_lines_style))
+            # Normal errorbars for points without upper/lower limits.
+            nolims = ~(lolims | uplims)
+            if nolims.any() and capsize > 0:
+                io, lo, ho = apply_mask([indep, low, high], nolims & everymask)
+                for lho in [lo, ho]:
+                    line = mlines.Line2D(io, io, marker=marker, **eb_cap_style)
+                    line.set(**{f"{name}data": lho})
+                    caplines.append(line)
```

Comment:
This is deeply confusing. At first, I thought `Lines2D(io, io, ...)` was a copy-paste error.

This construct at least needs a comment, e.g. 

> Since this has to work for x and y as dependent data, we first
> set both x and y to the independent variable and overwrite the
> respective dependent data afterwards.


More general, I'm not convinced this indirection is worth the trouble. Explicit seems better here.
```suggestion
                if name == 'x':
                    caplines.extend([
                        mlines.Line2D(lo, io, marker=marker, **eb_cap_style),
                        mlines.Line2D(ho, io, marker=marker, **eb_cap_style)])
                else:
                    caplines.extend([
                        mlines.Line2D(io, lo, marker=marker, **eb_cap_style),
                        mlines.Line2D(io, ho, marker=marker, **eb_cap_style)])
```
(Please check that I got all the `*o` right.)

Reply:
I went with the comment.